### ISSUE TITLE
Cleaner IdCitation extents in find_citations.py

### ISSUE THREAD

**Issue by slbayer** [freelawproject/courtlistener#1344]
As I described in issue 1338, I'm using a portion of the CourtListener code to write a standalone citation finder. This component is an NLP mention finder, so the extent of the component turns out to be important. I know that this isn't the intention of the citation finder in CourtListener, but in some cases the goals may overlap. The extend of the IdCitation may be one of these cases. 

By default, in cl/citations/find_citations.py, the IdCitation is created with a fixed chunk of subsequent tokens:

```
            citation = IdCitation(
                id_token=citation_token, after_tokens=words[i + 1 : i + 3]
            )
```

If you care about the actual extent of the remainder of the IdCitation, we can do better. In the corpus of court rulings and party filings I'm working with, something like this appears to be cleaner (this is in Python 3, so there may be some minor syntactic differences):

```
# Unicode 00b6 is the paragraph symbol.
ID_REF_TOKEN_LITERALS = set(["at", "p.", "p", "pp.", "p", "@", "pg", "pg.",
                            '\u00b6', '\u00b6\u00b6'])

# I will currently miss (a), etc. as subparagraphs. 2013 and 2014 are en- and em-dash.
ID_REF_NUM_PAT = re.compile("^[*\u00b6]*[0-9:\-\u2013\u2014,]+[.,;]?$")

...
            idx = i + 1
            while idx < endIdx:
                tok = words[idx]
                if (tok not in ID_REF_TOKEN_LITERALS) and (ID_REF_NUM_PAT.match(tok) is None):
                    break
                idx += 1            
            citation = IdCitation(
                id_token=citation_token, after_tokens=words[i + 1:idx]
            )
```

The basic idea is that Id. can be followed by any of the tokens in ```ID_REF_TOKEN_LITERALS``` or be a token which optionally starts with a star or the paragraph symbol, and is then a number-ish thing, maybe with dashes, commas or colons in it, optionally followed by a punctuation element.

I don't have a test suite to rigorously test this, but I do have a collection of about 2000 documents, with about 25K Id. citations, and eyeballing the result looks pretty clean.

**Comment by mlissner** [comment]
@mattdahl, do you have any interest in looking at this one?

**Comment by mattdahl** [comment]
@mlissner Sure, I should have time for this!

@slbayer Thanks for sharing! Just so I can visualize what `ID_REF_TOKEN_LITERALS` and `ID_REF_NUM_PAT` apply to in practice, can you paste a few raw strings of the kinds of Id. citations this code catches?

**Comment by slbayer** [comment]
Here are a few examples of some of the odder things I've found:

```

Id. at 21–22

Id. ¶ 34.

Id. 51-52.

Id. at 62-63, 67-68.

Id. at *10.

Id. at 7-9, ¶¶ 38-53.

Id. at pp.  45, 64.

id. 119:12-14

```

I've pulled all these from judges' decisions; they tend to be more careful than the party attorneys; for instance, I've only seen an ```@``` for "at" in the party filings. 

I'm also wondering whether I've missed a more important generalization here: the suffix of Id. citations, in the citation guides I've seen on-line, appears to be identical to the suffix of shortform citations. And in reviewing my data for you, I now see that I have a handful of Ibid. references which could stand a cleaner treatment as well.

Finally, I should ask how interested you folks are in identifying the proper token extent of citations in general. There's a much, much larger "bug" from that point of view in your overall strategy; your main loop looks for "anchors" for citations in the token sequence and builds out to the left and right from those anchors, and the result can lead to overlapping citation extentions. I can file an issue about this if you're interested, but it may diverge enough from what you care about that I shouldn't bother. The fix I have is pretty complicated, and would involve a wholesale reorganization of the core of find_citations.py.
